### DISCUSSION TITLE
How to delete a zombie task

### DISCUSSION THREAD

**Discussion by lbj2004032** [filecoin-project/lotus#9809] [Storage Provider]
Many early encapsulation tasks have problems, but they cannot be deleted. Is there a good way to clean them up?

$lotus-miner info
Sectors:
        Total: 156447
        Available: 147857
        Committing: 147
        CommitFinalize: 300
        FinalizeSector: 21
        Removed: 8122

How to delete a zombie task

The above "Committing, CommitFinalize, FinalizeSector" tasks are dead. However, the command "lotus miner remove --really-do-it "cannot be used to delete

**Comment by LexLuthr** [comment]
Check if there are pending requests in `lotus-miner sealing sched-diag`
If there are requests for non-existing sectors, you can abort them with `lotus-miner sealing abort --sched <Request UUID>`

**Reply by lbj2004032** [reply]
> Check if there are pending requests in `lotus-miner sealing sched-diag` If there are requests for non-existing sectors, you can abort them with `lotus-miner sealing abort --sched <Request UUID>`

thank you ,i will try it

**Reply by lbj2004032** [reply]
> lotus-miner sealing abort --sched <Request UUID>

Sectors:
        Total: 137153
        Proving: 136556
        Available: 18
        : 1
        TerminateWait: 1
        Removed: 521


Really effective
But there is another case
How to handle this state?
: 1
TerminateWait: 1

![1670429975615](https://user-images.githubusercontent.com/101935091/206233311-19af1518-25a4-4c2e-96c2-abbf4d7a3503.png)

**Reply by LexLuthr** [reply]
Just flush the termination and that should process it.
`lotus-miner sectors terminate flush` 
If this doesn't work then check miner logs to figure out what process might be holding off the termination.

**Reply by lbj2004032** [reply]
> Just flush the termination and that should process it. `lotus-miner sectors terminate flush` If this doesn't work then check miner logs to figure out what process might be holding off the termination.

$ lotus-miner sectors terminate flush
ERROR: no sectors were queued for termination

**Reply by lbj2004032** [reply]
> If this doesn't work then check miner logs to figure out what process might be holding off the termination.

2022-11-30T18:38:57.583+0800    ERROR   sectors pipeline/fsm.go:37      unhandled sector error (11243): entered TerminateWait with nil TerminateMessage:

**Reply by lbj2004032** [reply]
> Just flush the termination and that should process it. `lotus-miner sectors terminate flush` If this doesn't work then check miner logs to figure out what process might be holding off the termination.

2022-12-01T02:15:34.918+0800    ERROR   sectors pipeline/fsm.go:37      unhandled sector error (2): entered TerminateWait with nil TerminateMessage:
    github.com/filecoin-project/lotus/storage/pipeline.(*Sealing).handleTerminateWait

**Reply by lbj2004032** [reply]
> Just flush the termination and that should process it. `lotus-miner sectors terminate flush` If this doesn't work then check miner logs to figure out what process might be holding off the termination.

What is the status of ": 1"?

**Reply by LexLuthr** [reply]
Looks like it is waiting for termination message. Can you check if msg has landed on chain or is stuck in mpool. If it has landed on chain then check your state to find that msg

**Reply by lbj2004032** [reply]
> Looks like it is waiting for termination message. Can you check if msg has landed on chain or is stuck in mpool. If it has landed on chain then check your state to find that msg

How to find?

**Reply by LexLuthr** [reply]
Give me your miner ID and the sector number

**Reply by lbj2004032** [reply]
> Give me your miner ID and the sector number

There are two different miner nodes
id: f01715688,     sector number:2
id: f01662849,     sector number:11243

**Reply by LexLuthr** [reply]
Please provide sector logs as well.

**Reply by lbj2004032** [reply]
> Please provide sector logs as well.

[11243.txt](https://github.com/filecoin-project/lotus/files/10207664/11243.txt)
[2.txt](https://github.com/filecoin-project/lotus/files/10207667/2.txt)

**Reply by LexLuthr** [reply]
You sector was terminated on-chain but I think FSM is still waiting for the message. I am assuming you don't have the old msgs from then. 

https://filfox.info/en/message/bafy2bzacebtr2xlprgkham6db3ufiozvmc5tgd7omr6u2z7rkbgyqqhmeudqe


We can try to force a state for FSM and see if it let's go of the waiting part.

Try `lotus-miner sectors update-state --really-do-it 11243 Removing`

**Reply by lbj2004032** [reply]
> lotus-miner sectors update-state --really-do-it 11243 Removing

Really deleted

But I don't understand. If you execute terminate and then remove, you should first remove and then terminate

Does the block become removed after terminate?

**Reply by LexLuthr** [reply]
If you want to get rid of a expired sector then you use "Remove", otherwise you should always "Terminate". The "Terminate" will remove sector from disk as well. There is no need to remove it manually. 

But on the other hand, if you "Remove" a live/faulty sector then it will not terminate on-chain and you will be penalised. 

This was a case of stuck FSM. I am glad that it was resolved.

**Reply by lbj2004032** [reply]
> lotus-miner sectors update-state --really-do-it 11243 Removing

Execute the commond "lotus-miner info"
Sectors:
        Total: 137153
        Proving: 136556
        Available: 18
        : 1
        Removed: 578

Workers: Seal(1) WdPoSt(0) WinPoSt(0)


An empty status is displayed as
:1

Is this a mistake? What kind of state is it

**Reply by LexLuthr** [reply]
The output of miner info has predefined states. One of sector probably is not matching any of those state. Check your `lotus-miner sectors list` to try to figure out which sector is not in good state.

**Reply by lbj2004032** [reply]
> The output of miner info has predefined states. One of sector probably is not matching any of those state. Check your `lotus-miner sectors list` to try to figure out which sector is not in good state.

$ lotus-miner info
Sectors:
        Total: 137153
        Proving: 136556
        Available: 18
        : 1
        Removed: 578

I calculated 3 blocks in all states, the total number is 137152

Proving: 136556

Available: 18

Removed: 578

can not find the block with empty status
